### PR TITLE
gh-157605: Accept empty Python frame stack in sampling frame chain parser

### PR DESCRIPTION
**Pull request by aidaodedjl**
## Description

A native thread can retain a Python thread state while waiting in C code with no Python functions executing (e.g. after `PyGILState_Ensure()` + `PyEval_SaveThread()`). The frame chain of that thread state holds only the `base_frame` sentinel. `process_frame_chain()` treated the resulting NULL first frame as a broken chain, raised `RuntimeError: Failed to parse initial frame in chain`, and failed the entire sampling call — discarding valid stacks from all other threads.

With this change, the unparseable-first-frame error is raised only when the first frame is **not** the `base_frame` sentinel. When it is the sentinel, the walk stops and the empty Python stack is accepted, so sampling returns results for the remaining threads.

The walk stops at the sentinel, so `last_frame_visited` still equals `base_frame_addr` and the complete-stack validation in the frame cache keeps working for the empty-stack case.

## Reproduction / evidence

- **Pre-fix (Linux, cross-process, issue reproducer):** `pin=1` (native thread retaining an empty-stack thread state) fails 40/40 samples with `RuntimeError: Failed to parse initial frame in chain`, with and without frame caching; `pin=0` baseline succeeds 20/20.
- **Post-fix (same environment):** all four `pin x cache_frames` combinations succeed 20/20 with the target thread's Python frames visible.

## Test

`test_external_inspection.py` gains a self-contained regression test (`TestEmptyPythonStackSampling`):

- compiles a small C helper at runtime that spawns a native thread retaining a Python thread state with an empty frame stack (loaded via `ctypes`);
- asserts same-process sampling succeeds and the target thread's Python frames stay visible, across `pin x cache_frames` combinations;
- skips gracefully when no C compiler is configured (including Windows, where `CC` is not set) and fails loudly if the helper itself fails to compile.

On this machine (macOS, 3.16 debug build) the test reproduces the failure before the fix (`{'failure': 5}`) and passes after (1.1 s).

* Issue: gh-157605

### CODE REVIEW

File: Modules/_remote_debugging/frames.c [hunk 1]
```diff
@@ -359,9 +359,16 @@ process_frame_chain(
         }
 
         if (frame == NULL && PyList_GET_SIZE(ctx->frame_info) == 0) {
-            const char *e = "Failed to parse initial frame in chain";
-            PyErr_SetString(PyExc_RuntimeError, e);
-            return -1;
+            if (frame_addr != ctx->base_frame_addr) {
+                const char *e = "Failed to parse initial frame in chain";
```
Comment (liuzhijie-0614) [inline Modules/_remote_debugging/frames.c:373]: How about the first frame is native frame?

**Comment by aidaodedjl** [comment]
Good catch — a NULL first frame with a valid continuation would have still failed the sample. The chain could legitimately start this way when the topmost frame object is transiently undecodable mid-update, or when the thread sits in a C call that owns no Python frame object; in that state the walk now treats the first frame exactly like a mid-chain native frame: it emits the `<native>` marker when native frames are enabled and is skipped otherwise.

The strict error now remains only for a dangling first frame with no continuation (`next == NULL`), which cannot be represented in the result — that preserves the old guard against a silently truncated chain, and it keeps the `Incomplete sample` validation semantics from the empty-stack case (a dangling first frame now ends the walk like the base_frame sentinel does, so threads whose current_frame is NULL keep returning an empty stack instead of regressing into an error).

I could not construct a reproducer for the native-first state in-process without racing the interpreter, so this path is covered by symmetry with the mid-chain handling rather than by a dedicated test — the empty-stack and regression tests still pass across in-tree, out-of-tree and cross-process configurations.

**Comment by aidaodedjl** [comment]
Correcting myself on the second half of the paragraph above: a dangling undecodable first frame (`next == NULL`) does not end the walk into an empty stack — it raises `Failed to parse initial frame in chain` right in that branch, which is the intended guard. And the `current_frame == NULL` mention was off too: that case never reaches this branch at all (the walk loop does not run), so its behavior is untouched by this PR. The only case that returns an empty stack is the base_frame sentinel as the first frame.

**Comment by liuzhijie-0614** [comment]
Good idea, only an incomplete chain has a NULL next_frame_addr (except at base_frame), right?

**Comment by aidaodedjl** [comment]
Right — that invariant is what makes the strict error safe. In a healthy chain the only frame whose `previous` is NULL is the base_frame sentinel itself (it is initialized that way in `pystate.c`, and an empty-stack thread has `current_frame == &base_frame`), so the unwinder only sees `next == NULL` on a first frame when the walk could not reach the sentinel: a torn read mid-update or memory corruption. Everything else with no continuation is precisely the sentinel and takes the `break` above. So the dangling-frame error stays as a detector for genuinely broken chains rather than a case the sampler is expected to hit in normal operation.

File: Modules/_remote_debugging/frames.c [hunk 2]
```diff
@@ -359,9 +359,25 @@ process_frame_chain(
...
-            return -1;
+            if (frame_addr == ctx->base_frame_addr) {
+                // A native thread that released the GIL with
+                // PyEval_SaveThread() keeps a thread state whose frame
+                // chain holds only the base_frame sentinel.  This is a
+                // valid empty Python stack: accept it and let other
+                // threads report their stacks.
+                break;
+            }
+            if (next_frame_addr == 0) {
```
Comment (liuzhijie-0614) [inline Modules/_remote_debugging/frames.c:378]: An invalid frame can wrongly enter the native-frame branch: it reads frame == NULL with next_frame_addr != 0 — the condition used to identify a native boundary — because a remote read is not an atomic snapshot and an invalidated frame can still retain a stale previous pointer. The code below then inserts a <native> frame and keeps walking the stale chain, producing many wrong frames. This happens often since the innermost frame is torn down on every return.  @pablogsal @aidaodedjl

Comment (liuzhijie-0614) [reply Modules/_remote_debugging/frames.c:378]: It might be necessary to identify in advance whether it is a native frame by using the "owner" field.

**Comment by aidaodedjl** [comment]
You're right, and thanks for the precise framing. Source-confirmed the mechanics: `clear_thread_frame()` drops the frame contents (`PyStackRef_CLEAR(frame->f_executable)` on top of `_PyFrame_ClearExceptCode()`) while `owner` and `previous` survive, so a returned-from frame reads back as `frame == NULL` with a non-NULL `previous` — and as you say, that is the steady state for the innermost frame between calls, not a rare race. With my previous patch such a first frame fell through to the mid-chain handling and would be reported as `<native>` followed by a stale chain.

The updated patch discriminates by `owner` as you suggested: a first frame with no code object and `owner != FRAME_OWNED_BY_INTERPRETER` is treated as a dropped frame — the walk stops and the thread reports an empty stack instead of a fabricated `<native>` boundary plus stale frames. Only an interpreter-owned frame with a continuation (a genuine C boundary, e.g. a thread sitting in a C call that owns no Python frame) keeps the `<native>` treatment. The walk-stop also sets a policy flag so the `Incomplete sample` validation doesn't reject these stacks.

Two things I want to flag rather than hide: (1) mid-chain dropped frames still get the `<native>` marker today — the same steady-state argument applies there (any sample whose walk passes a returned-from frame mid-chain), so that likely wants the same owner check; I kept this patch scoped to the first-frame case under discussion but can extend it if you agree. (2) the strict error for a dangling first frame is gone under the owner policy (an interpreter-owned frame with `next == NULL` is indistinguishable from the sentinel), which trades a bit of corruption detection for not failing samples in the common state — sounded right for a sampler, but it's your call.

**Comment by aidaodedjl** [comment]
Done — the owner field now gates both paths. `is_frame_valid()` reports the owner of a code-less frame back to the chain walk, and the walk stops there whenever the owner is not `FRAME_OWNED_BY_INTERPRETER`, reporting the frames collected so far: a dropped first frame yields an empty stack, and a dropped mid-chain frame ends the chain instead of fabricating a `<native>` boundary followed by stale frames. Only interpreter-owned frames (the sentinel and genuine native boundaries) keep the `<native>` treatment. The owner is always read from the local frame buffer — the walk never dereferences a remote address. Truncated stacks were already excluded from the frame cache by the `last_frame_visited` completeness check, so nothing stale gets cached.

Regression coverage: the fake-frame helper is parameterized by owner, and the test locks in both behaviors — interpreter-owned first frame: skipped without native frames, `<native>` with native frames enabled; dropped (THREAD-owned) first frame: empty stack with native frames enabled, which fails on the previous commit where it produced `<native>` plus a stale walk. Verified in-tree and in a separate out-of-tree build.

**Comment by liuzhijie-0614** [comment]
> You're right, and thanks for the precise framing. Source-confirmed the mechanics: `clear_thread_frame()` drops the frame contents (`PyStackRef_CLEAR(frame->f_executable)` on top of `_PyFrame_ClearExceptCode()`) while `owner` and `previous` survive, so a returned-from frame reads back as `frame == NULL` with a non-NULL `previous` — and as you say, that is the steady state for the innermost frame between calls, not a rare race. With my previous patch such a first frame fell through to the mid-chain handling and would be reported as `<native>` followed by a stale chain.
> 
> The updated patch discriminates by `owner` as you suggested: a first frame with no code object and `owner != FRAME_OWNED_BY_INTERPRETER` is treated as a dropped frame — the walk stops and the thread reports an empty stack instead of a fabricated `<native>` boundary plus stale frames. Only an interpreter-owned frame with a continuation (a genuine C boundary, e.g. a thread sitting in a C call that owns no Python frame) keeps the `<native>` treatment. The walk-stop also sets a policy flag so the `Incomplete sample` validation doesn't reject these stacks.
> 
> Two things I want to flag rather than hide: (1) mid-chain dropped frames still get the `<native>` marker today — the same steady-state argument applies there (any sample whose walk passes a returned-from frame mid-chain), so that likely wants the same owner check; I kept this patch scoped to the first-frame case under discussion but can extend it if you agree. (2) the strict error for a dangling first frame is gone under the owner policy (an interpreter-owned frame with `next == NULL` is indistinguishable from the sentinel), which trades a bit of corruption detection for not failing samples in the common state — sounded right for a sampler, but it's your call.

Thanks! The solution I was thinking of:

1. Keep the handling of a dangling first frame unchanged — still require frame == NULL and next_frame_addr == 0.
2. But the native-frame detection below should be based on the owner. If the frame is being cleared or is otherwise invalid, it should not be added to frame_info, while the sampler should keep running normally.

If we only change (1) but not (2), a lot of spurious native frames show up — I've confirmed this in local experiments.

**Comment by aidaodedjl** [comment]
Landed in 2c58f57, following your scheme exactly: dangling code-less first frames keep raising `Failed to parse initial frame in chain`; native detection is owner-based, so a dropped frame (owner THREAD/GENERATOR) is never added to `frame_info` and the walk continues through its `previous` pointer to the live callers; the sentinel still terminates the walk by address and yields an empty stack. Mid-chain dropped frames get the same treatment, which removes the spurious `<native>` bursts you confirmed locally.

One subtle point from testing: the skip path re-assigns `prev_frame_addr` before continuing, otherwise the chain-consistency check mis-fires on the next real frame — worth a look since you've been inside this code recently. The fake-frame test helper is parameterized by owner now, so both the `<native>` boundary case and the skip-past-dropped-frame case are locked in, with and without frame caching. Verified in-tree and in a separate out-of-tree build; full `test_external_inspection` + `test_profiling` pass.

File: Modules/_remote_debugging/frames.c [hunk 3]
```diff
@@ -359,20 +367,54 @@ process_frame_chain(
...
+                PyErr_SetString(PyExc_RuntimeError, e);
+                return -1;
+            }
+            // Any other code-less first frame with a continuation is
+            // handled by the owner check below: an interpreter-owned
+            // frame is a native boundary, a dropped frame (owner
+            // THREAD/GENERATOR) is skipped, and the walk continues.
         }
         PyObject *extra_frame = NULL;
+        if (frame == NULL && frame_owner != FRAME_OWNED_BY_INTERPRETER) {
```
Comment (liuzhijie-0614) [inline Modules/_remote_debugging/frames.c:394]: This "if" can be removed?

**Comment by pablogsal** [comment]
I'm closing this. Please read https://devguide.python.org/getting-started/ai-tools/
